src/components/codeTabs.tsx

```diff
@@ -20,51 +22,69 @@ const LANGUAGES = {
+  children: React.ReactElement<CodeBlockProps> | React.ReactElement<CodeBlockProps>[];
+}
 
-export function CodeTabs({children}: Props) {
-  if (!Array.isArray(children)) {
-    children = [children];
-  } else {
-    children = [...children];
-  }
+export function CodeTabs({children}: CodeTabProps) {
+  const codeBlocks = Array.isArray(children) ? [...children] : [children];
 
   // the idea here is that we have two selection states.  The shared selection
   // always wins unless what is in the shared selection does not exist on the
-  // individual code block.  In that case the local selection overrides.  The
+  // individual code block. In that case the local selection overrides.  The
   // final selection is what is then rendered.
 
   const codeContext = useContext(CodeContext);
-  const [sharedSelection, setSharedSelection] = codeContext.sharedCodeSelection;
-  const [localSelection, setLocalSelection] = useState(null);
-  const [lastScrollOffset, setLastScrollOffset] = useState(null);
-  const tabBarRef = useRef(null);
+  const [localSelection, setLocalSelection] = useState<string | null>(null);
+  const [lastScrollOffset, setLastScrollOffset] = useState<number | null>(null);
+  const containerRef = useRef<HTMLDivElement>(null);
+
+  // When the selection switches we scroll so that the box that was toggled
+  // stays scrolled like it was before. This is because boxes above the changed
+  // box might also toggle and change height.
+  useEffect(() => {
+    if (containerRef.current === null) {
+      return;
+    }
+
+    if (lastScrollOffset !== null) {
+      const diff = containerRef.current.getBoundingClientRect().y - lastScrollOffset;
+      window.scroll(window.scrollX, window.scrollY + diff);
+      setLastScrollOffset(null);
+    }
+  }, [lastScrollOffset]);
 
   // The title is what we use for sorting and also for remembering the
-  // selection.  If there is no title fall back to the title cased language
-  // name (or override from `LANGUAGES`).
-  let possibleChoices = children.map(x => {
-    const {title, language} = x.props;
-    return (
-      title ||
-      LANGUAGES[language] ||
-      (language ? language[0].toUpperCase() + language.substring(1) : 'Text')
-    );
+  // selection. If there is no title fall back to the title cased language name
+  // (or override from `LANGUAGES`).
+  const possibleChoices = codeBlocks.map<string>(({props: {title, language}}) => {
+    if (title) {
+      return title;
+    }
+    if (!language) {
+      return 'Text';
+    }
+    if (language in HUMAN_LANGUAGE_NAMES) {
+      return HUMAN_LANGUAGE_NAMES[language];
+    }
+
+    return language[0].toUpperCase() + language.substring(1);
   });
 
   // disambiguate duplicates by enumerating them.
-  const tabTitleSeen = {};
-  possibleChoices = possibleChoices.reduce((arr, tabTitle) => {
-    if (possibleChoices.filter(x => x === tabTitle).length > 1) {
-      const num = (tabTitleSeen[tabTitle] = (tabTitleSeen[tabTitle] || 0) + 1);
```

Comment:
Tried to make this a bit less crazy as well, though `??=` may not be clear to everyone

Reply:
Yeah, and operators can be hard to google. You could add a comment with the operator name (Nullish coalescing assignment, I think)